crates/ruff/src/rules/ruff/rules/mutable_defaults_in_dataclass_fields.rs

```diff
@@ -0,0 +1,207 @@
+///
+/// @dataclass
+/// class B:
+///     also_mutable_default_but_sneakier: A = field(default_factory=A)
+/// ```
+///
+/// Alternatively, if you __want__ the shared behaviour, make it more obvious
+/// by assigning it to a module-level variable:
+/// ```python
+/// from dataclasses import dataclass
+///
+/// def creating_list() -> list[]:
+///     return [1, 2, 3, 4]
+///
+/// I_KNOW_THIS_IS_SHARED_STATE = creating_list()
+///
+/// @dataclass
+/// class A:
+///     mutable_default: list[int] = I_KNOW_THIS_IS_SHARED_STATE
+/// ```
+#[violation]
+pub struct FunctionCallInDataclassDefaultArgument {
+    pub name: Option<String>,
+}
+
+impl Violation for FunctionCallInDataclassDefaultArgument {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let FunctionCallInDataclassDefaultArgument { name } = self;
+        if let Some(name) = name {
+            format!("Do not perform function call `{name}` in dataclass defaults")
+        } else {
+            format!("Do not perform function call in dataclass defaults")
+        }
+    }
+}
+
+fn is_mutable_expr(expr: &Expr) -> bool {
+    matches!(
+        &expr.node,
+        ExprKind::List { .. }
+            | ExprKind::Dict { .. }
+            | ExprKind::Set { .. }
+            | ExprKind::ListComp { .. }
+            | ExprKind::DictComp { .. }
+            | ExprKind::SetComp { .. }
+    )
+}
+
+const ALLOWED_FUNCS: &[&[&str]] = &[&["dataclasses", "field"]];
+
+fn is_allowed_func(checker: &Checker, func: &Expr) -> bool {
+    checker
+        .ctx
+        .resolve_call_path(func)
+        .map_or(false, |call_path| {
+            ALLOWED_FUNCS
+                .iter()
+                .any(|target| call_path.as_slice() == *target)
+        })
+}
+
+pub fn function_call_in_dataclass_defaults(checker: &mut Checker, body: &[Stmt]) {
+    for statement in body {
+        if let StmtKind::AnnAssign {
+            value: Some(expr), ..
+        } = &statement.node
+        {
+            if let ExprKind::Call { func, .. } = &expr.node {
+                if !is_allowed_func(checker, func) {
+                    checker.diagnostics.push(Diagnostic::new(
+                        FunctionCallInDataclassDefaultArgument {
+                            name: compose_call_path(func),
+                        },
+                        Range::from(expr),
+                    ));
+                }
+            }
+        }
+    }
+}
+
+pub fn mutable_dataclass_default(checker: &mut Checker, body: &[Stmt]) {
+    for statement in body {
+        match &statement.node {
+            StmtKind::AnnAssign {
+                value: Some(value), ..
+            }
+            | StmtKind::Assign { value, .. } => {
+                if is_mutable_expr(value) {
+                    checker
+                        .diagnostics
+                        .push(Diagnostic::new(MutableDataclassDefault, Range::from(value)));
+                }
+            }
+            _ => (),
```

Comment:
Since there's only one case here, you can write this as an `if let`:

```rust
pub fn mutable_dataclass_default(checker: &mut Checker, body: &[Stmt]) {
    for statement in body {
        if let StmtKind::Assign { value, .. }
        | StmtKind::AnnAssign {
            value: Some(value), ..
        } = &statement.node
        {
            if is_mutable_expr(value) {
                checker
                    .diagnostics
                    .push(Diagnostic::new(MutableDataclassDefault, Range::from(value)));
            }
        }
    }
}
```

Reply:
I actually experimented quite a bit, the syntax is still quite new to me. Somehow I did completely miss doing this, thanks!